### ISSUE TITLE
Dragonfly performance issues, misconfiguration? (help wanted)

### ISSUE THREAD

**Issue by ish-xyz** [dragonflyoss/Dragonfly2#692]
I'm trying to run Dragonfly on a small cluster with 6 nodes and I'm noticing some performance issues when I run 1000 docker pulls at the same time.
I'm using a tool that I developed to perform the docker pulls it just triggers multiple go routines and each go routine will ask for the docker image manifest and download the layers.
If I trigger 100/200 pulls in parallel everything it's fine, as soon as I increase to 800/1000 I start noticing 500 errors in the requests that the tool does.

Currently I have 3 schedulers, 6 dfdaemons (one for each node), 1 cdn (deployed as statefulset).
Each node in the cluster have 65GB ram and there's a 1GB/s connectivity between nodes.

The configurations used are the following:

scheduler:
```
 scheduler:
    workerNum: 1000000
    backSourceCount: 3
    accessWindow: 3m
    candidateParentCount: 100
    cdnLoad: 10000
    clientLoad: 1000

  dynconfig:
      cdnDirPath: /opt/dragonfly/scheduler-cdn

  disableCDN: false
  console: true
  verbose: false
```

cdn:
```
    base:
      listenPort: 8003
      downloadPort: 8001
      systemReservedBandwidth: 20M
      maxBandwidth: 1024M
      enableProfiler: false
      failAccessInterval: 3m
      gcInitialDelay: 6s
      gcMetaInterval: 2m
      gcStorageInterval: 15s
      taskExpireTime: 3m
      storagePattern: disk
      console: true
    plugins:
      storageDriver:
      - name: disk
        enable: true
        config:
          baseDir: /var/lib/dragonfly/cdn
      storageManager:
      - name: disk
        enable: true
        config:
          gcInitialDelay: 5s
          gcInterval: 15s
          driverConfigs:
            disk:
              gcConfig:
                youngGCThreshold: 100G
                fullGCThreshold: 5G
                cleanRatio: 1
                intervalThreshold: 2h
```

dfdaemon:
```
   # daemon alive time, when sets 0s, daemon will not auto exit
    # it is useful for longtime running
    aliveTime: 0s
    console: true
    # when enable, pprof will be enabled
    verbose: false
    gcInterval: 1m0s
    dataDir: /var/lib/dragonfly
    workHome: /var/lib/dragonfly
    keepStorage: false

    scheduler:
      scheduleTimeout: 30s
      # when true, only scheduler says back source, daemon can back source
      disableAutoBackSource: true
      netAddrs:
        - type: tcp
          addr: scheduler-0:8002
        - type: tcp
          addr: scheduler-1:8002
        - type: tcp
          addr: scheduler-2:8002


    # current host info used for scheduler
    host:
      listenIP: 0.0.0.0
      advertiseIP: 0.0.0.0
      location: {{ .Values.clusterName }}
      idc: {{ .Values.datacenter }}
      securityDomain: {{ .Values.datacenter }}
      netTopology: ""

    # download service option
    download:
      totalRateLimit: 500Mi
      perPeerRateLimit: 250Mi
      pieceDownloadTimeout: 60s
      downloadGRPC:
        # security option
        security:
          insecure: true #TODO
          cacert: ""
          cert: ""
          key: ""
          tlsConfig: null
        unixListen:
          socket: /var/run/dfdaemon.sock
      # peer grpc option
      # peer grpc service send pieces info to other peers
      peerGRPC:
        security:
          insecure: true
          cacert: ""
          cert: ""
          key: ""
        tcpListen:
          # listen address
          listen: 0.0.0.0
          port: 65000

    # upload service option
    upload:
      # upload limit per second
      rateLimit: 500Mi
      security:
        insecure: true
        cacert: ""
        cert: ""
        key: ""
      tcpListen:
        # listen address
        listen: 0.0.0.0
        port: 65002

    # peer task storage option
    storage:
      # task data expire time
      # when there is no access to a task data, this task will be gc.
      taskExpireTime: 3m0s
      strategy: io.d7y.storage.v2.advance
      diskGCThreshold: {{ .Values.dragonfly.dfdaemon.diskGCThreshold }} #50Gi by default
      # set to ture for reusing underlying storage for same task id
      multiplex: true

    proxy:
      defaultFilter: "Expires&Signature"
      security:
        insecure: true
        cacert: ""
        cert: ""
        key: ""
      tcpListen:
        namespace: ""
        listen: 0.0.0.0
        port: 65001

      registryMirror:
        url: http://registry:8080
        insecure: true
        certs: []
        direct: false
```

When I notice the error (500 error) the logs in the scheduler says:

```
2021-09-27T13:59:13.548Z        DEBUG   basic/basic_scheduler.go:264    finish schedule parent flow, select num 0 candidates parents, current task tree node count 1, back source peers: [10.1.75.159-1-6ef75e8a-8edc-48bd-bdba-d0cb8d125cfb 10.1.247.1-1-ff71e57e-3a2d-4684-9e66-0fa0ff1dc323 10.1.160.193-1-2efed1a1-1cf3-45d0-b0d4-b618e5e5c9b4 10.1.160.193-1-d925f7ee-d5c9-4434-b786-48cfaa9d3fb3 10.1.0.94-1-11f6b3c9-0f23-4173-9d55-4ad7455bd64f 10.1.160.193-1-05c821f3-4ed0-4e32-9a5c-b569e5d402d1 10.1.247.1-1-19cd5d9f-535e-4fc2-b4a4-b796bededf66 10.1.75.159-1-73c47102-3ca5-48a7-a9a2-fccdc83eb47d 10.1.160.193-1-7993e2b0-4856-45b2-8253-15247bddfc99 10.1.247.1-1-dfe03494-df37-4d6f-a961-1350472b220b 10.1.210.56-1-4e04fc36-a545-4bca-a38f-cea685ac0b72 10.1.48.13-1-11c84096-a2e1-4ebc-a387-b4919b2b09ef 10.1.75.159-1-0c9f3acc-f95a-4d71-b443-c16071a759dc  .....(VERY LONG LIST) ........
```

```
xx@xx:~/xx/xx$ k logs scheduler-0 | grep ERROR
2021-09-27T13:59:22.451Z        ERROR   core/events.go:94       reScheduleParent: failed to schedule parent to peer 10.1.160.193-1-a19c2466-cfb6-4e17-8cab-a8a45d98456d, reschedule it later
2021-09-27T13:59:22.468Z        ERROR   core/events.go:94       reScheduleParent: failed to schedule parent to peer 10.1.160.193-1-7aa1054a-4518-48c9-96cc-42de32901d65, reschedule it later
2021-09-27T13:59:22.473Z        ERROR   core/events.go:94       reScheduleParent: failed to schedule parent to peer 10.1.0.94-1-81e52445-af49-4764-b38b-984a43180d38, reschedule it later
2021-09-27T13:59:22.505Z        ERROR   core/events.go:94       reScheduleParent: failed to schedule parent to peer 10.1.75.159-1-ff2fc174-6460-4808-bcab-f7fc291d5f41, reschedule it later
2021-09-27T13:59:23.354Z        ERROR   core/events.go:94       reScheduleParent: failed to schedule parent to peer 10.1.48.13-1-b7e7bcb7-3678-4057-809d-7fd4fe4c5d28, reschedule it later
2021-09-27T13:59:23.418Z        ERROR   core/events.go:94       reScheduleParent: failed to schedule parent to peer 10.1.75.159-1-a5d4a2ba-7d9b-492c-b8cd-837462e5ebca, reschedule it later
(VERY LONG LIST)
```

What am I missing? Can you spot any clear error @jim3ma?

I've disabled autobacksource on purpose as I don't want requests going to the upstream from the dfdaemons.

**Comment by jim3ma** [comment]
1. Do you pull same images or not ? 
2. What is the size of the images ?
3. `grep 10.1.75.159-1-6ef75e8a-8edc-48bd-bdba-d0cb8d125cfb [dfdaemon log]`, and paste here.

**Comment by ish-xyz** [comment]
Here's the log:

```
ishama@czwishama1:~/repos/dragonfly$ k logs -f dfdaemon-z8kwc | grep 10.1.247.2-1-a2d7ec6d-7e9a-4d22-9f44-4a6e66a62d1e
2021-09-27T14:32:21.709Z        INFO    transport/transport.go:159      start download with url: https://docker-remote.artifactory.c3.zone/v2/debian/blobs/sha256:955615a668ce169f8a1443fc6b6e6215f43fe0babfb4790712a2d3171f34d366      {"peer": "10.1.247.2-1-a2d7ec6d-7e9a-4d22-9f44-4a6e66a62d1e", "component": "transport"}
2021-09-27T14:32:21.709Z        DEBUG   peer/peertask_stream.go:71      request overview, pid: 10.1.247.2-1-a2d7ec6d-7e9a-4d22-9f44-4a6e66a62d1e, url: https://docker-remote.artifactory.c3.zone/v2/debian/blobs/sha256:955615a668ce169f8a1443fc6b6e6215f43fe0babfb4790712a2d3171f34d366, filter: Expires&Signature, meta: tag:"d7y/proxy" filter:"Expires&Signature" header:{key:"Accept" value:"application/vnd.docker.distribution.manifest.v2+json"} header:{key:"Authorization" value:"Basic c3ZjLWs4cy1hcnRpZmFjdG9yeTpYP1VSVTBhZEt1ZEY4QVUxZHp5TT9GQTZn"} header:{key:"User-Agent" value:"Go-http-client/1.1"} header:{key:"X-Forwarded-For" value:"10.1.160.0"}, tag: d7y/proxy
2021-09-27T14:32:21.709Z        INFO    peer/peertask_stream.go:77      step 1: peer 10.1.247.2-1-a2d7ec6d-7e9a-4d22-9f44-4a6e66a62d1e start to register
2021-09-27T14:32:21.709Z        INFO    client/client.go:83     generate hash key taskId: a89c50d7acc685dca624cd85f7c7e4f0938da9dc0ffe62c8d813162401f1a27b and start to register peer task for peer_id(10.1.247.2-1-a2d7ec6d-7e9a-4d22-9f44-4a6e66a62d1e) url(https://docker-remote.artifactory.c3.zone/v2/debian/blobs/sha256:955615a668ce169f8a1443fc6b6e6215f43fe0babfb4790712a2d3171f34d366)        {"taskId": "a89c50d7acc685dca624cd85f7c7e4f0938da9dc0ffe62c8d813162401f1a27b", "peerID": "10.1.247.2-1-a2d7ec6d-7e9a-4d22-9f44-4a6e66a62d1e"}
2021-09-27T14:32:21.710Z        INFO    client/client.go:106    register peer task result success url: https://docker-remote.artifactory.c3.zone/v2/debian/blobs/sha256:955615a668ce169f8a1443fc6b6e6215f43fe0babfb4790712a2d3171f34d366, scheduler: dns:///scheduler-0.scheduler-svc:8002      {"taskId": "a89c50d7acc685dca624cd85f7c7e4f0938da9dc0ffe62c8d813162401f1a27b", "peerID": "10.1.247.2-1-a2d7ec6d-7e9a-4d22-9f44-4a6e66a62d1e"}
2021-09-27T14:32:21.710Z        INFO    peer/peertask_stream.go:108     register task success, task id: a89c50d7acc685dca624cd85f7c7e4f0938da9dc0ffe62c8d813162401f1a27b, peer id: 10.1.247.2-1-a2d7ec6d-7e9a-4d22-9f44-4a6e66a62d1e, SizeScope: NORMAL
2021-09-27T14:32:21.710Z        DEBUG   peer/peertask_stream.go:137     a89c50d7acc685dca624cd85f7c7e4f0938da9dc0ffe62c8d813162401f1a27b/10.1.247.2-1-a2d7ec6d-7e9a-4d22-9f44-4a6e66a62d1e size scope: normal
2021-09-27T14:32:21.710Z        INFO    client/client.go:154    start to report piece result for taskID: a89c50d7acc685dca624cd85f7c7e4f0938da9dc0ffe62c8d813162401f1a27b       {"peerId": "10.1.247.2-1-a2d7ec6d-7e9a-4d22-9f44-4a6e66a62d1e", "errMsg": null}
2021-09-27T14:32:21.711Z        INFO    peer/peertask_stream.go:142     step 2: start report peer 10.1.247.2-1-a2d7ec6d-7e9a-4d22-9f44-4a6e66a62d1e piece result
2021-09-27T14:32:31.718Z        ERROR   scheduler/scheduler_grpc.pb.go:75       client receive a message: *scheduler.PeerPacket error: [5001]reschedule parent for peer 10.1.247.2-1-a2d7ec6d-7e9a-4d22-9f44-4a6e66a62d1e already reaches max reschedule times for method: /scheduler.Scheduler/ReportPieceResult target: dns:///scheduler-0.scheduler-svc:8002 connState: READY
2021-09-27T14:32:31.718Z        ERROR   peer/peertask_base.go:513       download from source disabled, auto back source disabled        {"peer": "10.1.247.2-1-a2d7ec6d-7e9a-4d22-9f44-4a6e66a62d1e", "task": "a89c50d7acc685dca624cd85f7c7e4f0938da9dc0ffe62c8d813162401f1a27b", "component": "streamPeerTask"}
2021-09-27T14:32:31.718Z        ERROR   peer/peertask_base.go:378       wait first peer packet error    {"peer": "10.1.247.2-1-a2d7ec6d-7e9a-4d22-9f44-4a6e66a62d1e", "task": "a89c50d7acc685dca624cd85f7c7e4f0938da9dc0ffe62c8d813162401f1a27b", "component": "streamPeerTask"}
2021-09-27T14:32:31.718Z        ERROR   peer/peertask_stream.go:398     end piece result sent, peer task failed {"peer": "10.1.247.2-1-a2d7ec6d-7e9a-4d22-9f44-4a6e66a62d1e", "task": "a89c50d7acc685dca624cd85f7c7e4f0938da9dc0ffe62c8d813162401f1a27b", "component": "streamPeerTask"}
2021-09-27T14:32:31.718Z        ERROR   peer/peertask_stream_callback.go:121    stream peer task failed, code: 1500, reason: download from source disabled      {"peer": "10.1.247.2-1-a2d7ec6d-7e9a-4d22-9f44-4a6e66a62d1e", "task": "a89c50d7acc685dca624cd85f7c7e4f0938da9dc0ffe62c8d813162401f1a27b", "component": "streamPeerTask"}
2021-09-27T14:32:31.719Z        INFO    peer/peertask_stream_callback.go:142    step 3: report fail peer result ok      {"peer": "10.1.247.2-1-a2d7ec6d-7e9a-4d22-9f44-4a6e66a62d1e", "task": "a89c50d7acc685dca624cd85f7c7e4f0938da9dc0ffe62c8d813162401f1a27b", "component": "streamPeerTask"}
2021-09-27T14:32:31.719Z        ERROR   peer/peertask_stream.go:262     download from source disabled   {"peer": "10.1.247.2-1-a2d7ec6d-7e9a-4d22-9f44-4a6e66a62d1e", "task": "a89c50d7acc685dca624cd85f7c7e4f0938da9dc0ffe62c8d813162401f1a27b", "component": "streamPeerTask"}
2021-09-27T14:32:31.719Z        ERROR   transport/transport.go:193      download fail: download from source disabled    {"peer": "10.1.247.2-1-a2d7ec6d-7e9a-4d22-9f44-4a6e66a62d1e", "component": "transport"}
```

I pull the same set of images (it's 4 diff images) the images are pretty small:

```
images:
- busybox:latest
- docker:latest
- debian:latest
- ubuntu:latest
```

**Comment by jim3ma** [comment]
@gaius-qi @244372610 Can you help @ish-xyz to debug the schedule failed ?

**Comment by jim3ma** [comment]
```
2021-09-27T14:32:31.718Z        ERROR   scheduler/scheduler_grpc.pb.go:75       client receive a message: *scheduler.PeerPacket error: [5001]reschedule parent for peer 10.1.247.2-1-a2d7ec6d-7e9a-4d22-9f44-4a6e66a62d1e already reaches max reschedule times for method: /scheduler.Scheduler/ReportPieceResult target: dns:///scheduler-0.scheduler-svc:8002 connState: READY
```

`5001` stands that scheduler says dfdaemon should download from source. You can disable back source and the `5001` is also be skipped. Maybe we need more precise option to back source when scheduler says `5001`

**Comment by ish-xyz** [comment]
some other logs from the scheduler
```
2021-09-27T15:30:10.949Z        ERROR   core/events.go:94       reScheduleParent: failed to schedule parent to peer 10.1.0.165-1-64790438-7b39-421b-8f3d-55f9a5e22eb8, reschedule it later
2021-09-27T15:30:10.949Z        DEBUG   basic/basic_scheduler.go:209    start schedule parent flow      {"taskId": "56499080341764aa9cdc67f0bad166d1f81239fbd37bdea0f80618fcec6a3210", "peerID": "10.1.247.4-1-a0273d88-8c3e-4baf-b32e-1483285f08e4"}
2021-09-27T15:30:10.949Z        DEBUG   basic/basic_scheduler.go:247    ++++++candidate parent peer cdn-0-56499080341764aa9cdc67f0bad166d1f81239fbd37bdea0f80618fcec6a3210_CDN is not selected because it's free upload load equal to less than zero++++++      {"taskId": "56499080341764aa9cdc67f0bad166d1f81239fbd37bdea0f80618fcec6a3210", "peerID": "10.1.247.4-1-a0273d88-8c3e-4baf-b32e-1483285f08e4"}
2021-09-27T15:30:10.949Z        DEBUG   basic/basic_scheduler.go:264    finish schedule parent flow, select num 0 candidates parents, current task tree node count 1, back source peers: []     {"taskId": "56499080341764aa9cdc67f0bad166d1f81239fbd37bdea0f80618fcec6a3210", "peerID": "10.1.247.4-1-a0273d88-8c3e-4baf-b32e-1483285f08e4"}
2021-09-27T15:30:10.949Z        ERROR   core/events.go:94       reScheduleParent: failed to schedule parent to peer 10.1.247.4-1-a0273d88-8c3e-4baf-b32e-1483285f08e4, reschedule it later
2021-09-27T15:30:10.950Z        DEBUG   basic/basic_scheduler.go:209    start schedule parent flow      {"taskId": "56499080341764aa9cdc67f0bad166d1f81239fbd37bdea0f80618fcec6a3210", "peerID": "10.1.75.167-1-ff2a2502-9ebc-48ba-ab1d-bd2a5902a590"}
2021-09-27T15:30:10.950Z        DEBUG   basic/basic_scheduler.go:247    ++++++candidate parent peer cdn-0-56499080341764aa9cdc67f0bad166d1f81239fbd37bdea0f80618fcec6a3210_CDN is not selected because it's free upload load equal to less than zero++++++      {"taskId": "56499080341764aa9cdc67f0bad166d1f81239fbd37bdea0f80618fcec6a3210", "peerID": "10.1.75.167-1-ff2a2502-9ebc-48ba-ab1d-bd2a5902a590"}
2021-09-27T15:30:10.950Z        DEBUG   basic/basic_scheduler.go:264    finish schedule parent flow, select num 0 candidates parents, current task tree node count 1, back source peers: []     {"taskId": "56499080341764aa9cdc67f0bad166d1f81239fbd37bdea0f80618fcec6a3210", "peerID": "10.1.75.167-1-ff2a2502-9ebc-48ba-ab1d-bd2a5902a590"}
2021-09-27T15:30:10.950Z        ERROR   core/events.go:94       reScheduleParent: failed to schedule parent to peer 10.1.75.167-1-ff2a2502-9ebc-48ba-ab1d-bd2a5902a590, reschedule it later
2021-09-27T15:30:10.958Z        INFO    rpcserver/rpcserver.go:151      peer 10.1.0.165-1-2b75232b-346c-4409-b470-5414cc34a664 is disconnect: [5001]reschedule parent for peer 10.1.0.165-1-2b75232b-346c-4409-b470-5414cc34a664 already reaches max reschedule times
2021-09-27T15:30:10.958Z        ERROR   rpc/server.go:216       reschedule parent for peer 10.1.0.165-1-2b75232b-346c-4409-b470-5414cc34a664 already reaches max reschedule times
2021-09-27T15:30:10.958Z        ERROR   grpc@v1.39.0/server.go:1548     do stream server error: rpc error: code = Unknown desc = [5001]reschedule parent for peer 10.1.0.165-1-2b75232b-346c-4409-b470-5414cc34a664 already reaches max reschedule times for method: /scheduler.Scheduler/ReportPieceResult
2021-09-27T15:30:10.959Z        ERROR   scheduler/scheduler_grpc.pb.go:183      server receive a message: *scheduler.PieceResult error: rpc error: code = Canceled desc = context canceled for method: /scheduler.Scheduler/ReportPieceResult
2021-09-27T15:30:10.960Z        DEBUG   rpcserver/rpcserver.go:174      report peer result task_id:"56499080341764aa9cdc67f0bad166d1f81239fbd37bdea0f80618fcec6a3210"  peer_id:"10.1.0.165-1-2b75232b-346c-4409-b470-5414cc34a664"  src_ip:"10.1.0.165"  security_domain:"none"  idc:"none"  url:"http://xxxx/v2/docker-remote/docker/blobs/sha256:f7a4137012260e097fc6b4bf6d75ba3b193acb399fb1276687303dea63ba519c"  content_length:-1  cost:10009  code:1500  total_piece_count:-1
2021-09-27T15:30:10.960Z        INFO    scheduler/scheduler_grpc.pb.go:202      finish peer task        {"success": false, "peerID": "10.1.0.165-1-2b75232b-346c-4409-b470-5414cc34a664", "taskID": "56499080341764aa9cdc67f0bad166d1f81239fbd37bdea0f80618fcec6a3210", "URL": "http://xxxx/v2/docker-remote/docker/blobs/sha256:f7a4137012260e097fc6b4bf6d75ba3b193acb399fb1276687303dea63ba519c", "IDC": "none", "peerIP": "10.1.0.165", "securityDomain": "none", "schedulerIp": "10.1.0.93", "schedulerName": "scheduler-0", "contentLength": "-1.0B", "traffic": "0.0B", "cost": "10.009µs", "code": 1500}
```

**Comment by ish-xyz** [comment]
I also have a question @jim3ma @gaius-qi, is it normal that it 'creates' so many peers? Shouldn't re-use the existing ones? Or is it because I'm running pulls in parallel and a single peer in only able to handle a task at a time?

**Comment by jim3ma** [comment]
> I also have a question @jim3ma @gaius-qi, is it normal that it 'creates' so many peers? Shouldn't re-use the existing ones? Or is it because I'm running pulls in parallel and a single peer in only able to handle a task at a time?

One layer one peer task. Running peer tasks can not reuse directly due to underlay context (a download can be cancelled, the context cancelled.). When a peer task is done, same tasks in this dfdaemon(peer host) will reuse the underlay data.

**Comment by 244372610** [comment]
> some other logs from the scheduler
> 
> ```
> 2021-09-27T15:30:10.949Z        ERROR   core/events.go:94       reScheduleParent: failed to schedule parent to peer 10.1.0.165-1-64790438-7b39-421b-8f3d-55f9a5e22eb8, reschedule it later
> 2021-09-27T15:30:10.949Z        DEBUG   basic/basic_scheduler.go:209    start schedule parent flow      {"taskId": "56499080341764aa9cdc67f0bad166d1f81239fbd37bdea0f80618fcec6a3210", "peerID": "10.1.247.4-1-a0273d88-8c3e-4baf-b32e-1483285f08e4"}
> 2021-09-27T15:30:10.949Z        DEBUG   basic/basic_scheduler.go:247    ++++++candidate parent peer cdn-0-56499080341764aa9cdc67f0bad166d1f81239fbd37bdea0f80618fcec6a3210_CDN is not selected because it's free upload load equal to less than zero++++++      {"taskId": "56499080341764aa9cdc67f0bad166d1f81239fbd37bdea0f80618fcec6a3210", "peerID": "10.1.247.4-1-a0273d88-8c3e-4baf-b32e-1483285f08e4"}
> 2021-09-27T15:30:10.949Z        DEBUG   basic/basic_scheduler.go:264    finish schedule parent flow, select num 0 candidates parents, current task tree node count 1, back source peers: []     {"taskId": "56499080341764aa9cdc67f0bad166d1f81239fbd37bdea0f80618fcec6a3210", "peerID": "10.1.247.4-1-a0273d88-8c3e-4baf-b32e-1483285f08e4"}
> 2021-09-27T15:30:10.949Z        ERROR   core/events.go:94       reScheduleParent: failed to schedule parent to peer 10.1.247.4-1-a0273d88-8c3e-4baf-b32e-1483285f08e4, reschedule it later
> 2021-09-27T15:30:10.950Z        DEBUG   basic/basic_scheduler.go:209    start schedule parent flow      {"taskId": "56499080341764aa9cdc67f0bad166d1f81239fbd37bdea0f80618fcec6a3210", "peerID": "10.1.75.167-1-ff2a2502-9ebc-48ba-ab1d-bd2a5902a590"}
> 2021-09-27T15:30:10.950Z        DEBUG   basic/basic_scheduler.go:247    ++++++candidate parent peer cdn-0-56499080341764aa9cdc67f0bad166d1f81239fbd37bdea0f80618fcec6a3210_CDN is not selected because it's free upload load equal to less than zero++++++      {"taskId": "56499080341764aa9cdc67f0bad166d1f81239fbd37bdea0f80618fcec6a3210", "peerID": "10.1.75.167-1-ff2a2502-9ebc-48ba-ab1d-bd2a5902a590"}
> 2021-09-27T15:30:10.950Z        DEBUG   basic/basic_scheduler.go:264    finish schedule parent flow, select num 0 candidates parents, current task tree node count 1, back source peers: []     {"taskId": "56499080341764aa9cdc67f0bad166d1f81239fbd37bdea0f80618fcec6a3210", "peerID": "10.1.75.167-1-ff2a2502-9ebc-48ba-ab1d-bd2a5902a590"}
> 2021-09-27T15:30:10.950Z        ERROR   core/events.go:94       reScheduleParent: failed to schedule parent to peer 10.1.75.167-1-ff2a2502-9ebc-48ba-ab1d-bd2a5902a590, reschedule it later
> 2021-09-27T15:30:10.958Z        INFO    rpcserver/rpcserver.go:151      peer 10.1.0.165-1-2b75232b-346c-4409-b470-5414cc34a664 is disconnect: [5001]reschedule parent for peer 10.1.0.165-1-2b75232b-346c-4409-b470-5414cc34a664 already reaches max reschedule times
> 2021-09-27T15:30:10.958Z        ERROR   rpc/server.go:216       reschedule parent for peer 10.1.0.165-1-2b75232b-346c-4409-b470-5414cc34a664 already reaches max reschedule times
> 2021-09-27T15:30:10.958Z        ERROR   grpc@v1.39.0/server.go:1548     do stream server error: rpc error: code = Unknown desc = [5001]reschedule parent for peer 10.1.0.165-1-2b75232b-346c-4409-b470-5414cc34a664 already reaches max reschedule times for method: /scheduler.Scheduler/ReportPieceResult
> 2021-09-27T15:30:10.959Z        ERROR   scheduler/scheduler_grpc.pb.go:183      server receive a message: *scheduler.PieceResult error: rpc error: code = Canceled desc = context canceled for method: /scheduler.Scheduler/ReportPieceResult
> 2021-09-27T15:30:10.960Z        DEBUG   rpcserver/rpcserver.go:174      report peer result task_id:"56499080341764aa9cdc67f0bad166d1f81239fbd37bdea0f80618fcec6a3210"  peer_id:"10.1.0.165-1-2b75232b-346c-4409-b470-5414cc34a664"  src_ip:"10.1.0.165"  security_domain:"none"  idc:"none"  url:"http://xxxx/v2/docker-remote/docker/blobs/sha256:f7a4137012260e097fc6b4bf6d75ba3b193acb399fb1276687303dea63ba519c"  content_length:-1  cost:10009  code:1500  total_piece_count:-1
> 2021-09-27T15:30:10.960Z        INFO    scheduler/scheduler_grpc.pb.go:202      finish peer task        {"success": false, "peerID": "10.1.0.165-1-2b75232b-346c-4409-b470-5414cc34a664", "taskID": "56499080341764aa9cdc67f0bad166d1f81239fbd37bdea0f80618fcec6a3210", "URL": "http://xxxx/v2/docker-remote/docker/blobs/sha256:f7a4137012260e097fc6b4bf6d75ba3b193acb399fb1276687303dea63ba519c", "IDC": "none", "peerIP": "10.1.0.165", "securityDomain": "none", "schedulerIp": "10.1.0.93", "schedulerName": "scheduler-0", "contentLength": "-1.0B", "traffic": "0.0B", "cost": "10.009µs", "code": 1500}
> ```

this is because CDN upload quota has been used up， The client needs to go back to the source itself to complete the task。The logic of this place scheduler also needs to be optimized

**Comment by jim3ma** [comment]
@ish-xyz Fixed by #693
